### PR TITLE
Add support for building with Clang 19

### PR DESCRIPTION
**Pull request by MichaelSimons**
Backport of https://github.com/dotnet/runtime/pull/105141 to release/8.0
Upstreaming [.NET 8.0 Fedora 41 patch](https://src.fedoraproject.org/rpms/dotnet8.0/blob/rawhide/f/runtime-clang-19.patch)

This is required so that source-build can build/test on Fedora 41 (https://github.com/dotnet/source-build/issues/4593).

### CODE REVIEW

**Review by am11 — APPROVED** [review]
LGTM. Thank you! 👍

File: src/native/libs/CMakeLists.txt
Comment (jkoritzinsky) [inline src/native/libs/CMakeLists.txt:130]: Could we instead use CMake's ability to check if the compiler supports this flag instead of also adding the "ignore unknown flags" option?

Comment (am11) [reply src/native/libs/CMakeLists.txt:130]:
```suggestion
        add_compile_options(-Wno-c11-extensions)

        check_c_compiler_flag(-Wc11-extensions COMPILER_SUPPORTS_W_C11_EXTENSIONS)
        if (COMPILER_SUPPORTS_W_C11_EXTENSIONS)
            add_compile_options(-Wno-c11-extensions)
        endif()

```

File: src/native/libs/CMakeLists.txt
Comment (am11) [inline src/native/libs/CMakeLists.txt:132]:
```suggestion

        check_c_compiler_flag(-Wpre-c11-compat COMPILER_SUPPORTS_W_PRE_C11_COMPAT)
        if (COMPILER_SUPPORTS_W_PRE_C11_COMPAT)
            add_compile_options(-Wno-pre-c11-compat)
        endif()

```

**Comment by omajid** [comment]
I forgot to note it down in my original Fedora patch, but the changes to `src/native/libs/CMakeLists.txt` are from a different PR: https://github.com/dotnet/runtime/pull/101871/

**Review by jkoritzinsky — APPROVED** [review]

**Comment by omajid** [comment]
cc @dviererbe who had to use a similar fix on Ubuntu

**Comment by dviererbe** [comment]
> cc @dviererbe who had to use a similar fix on Ubuntu

Correct, we have a [patch](https://git.launchpad.net/ubuntu/+source/dotnet8/tree/debian/patches/0004-fix-clang19-build.patch?h=ubuntu%2Foracular) for Ubuntu >= 24.10. Ubuntu 24.04 uses clang 18 and Ubuntu 22.04 uses clang 14 and therefore does not require the patch.

Nice that this gets backported :)

**Comment by dviererbe** [comment]
We also have a [patch](https://git.launchpad.net/ubuntu/+source/dotnet8/tree/debian/patches/0001-fix-clang18-build.patch?h=ubuntu/oracular) in place for clang 18. Shouldn't this be backported first before adding clang 19 support?

**Review by jkoritzinsky — APPROVED** [review]

**Comment by MichaelSimons** [comment]
> We also have a [patch](https://git.launchpad.net/ubuntu/+source/dotnet8/tree/debian/patches/0001-fix-clang18-build.patch?h=ubuntu/oracular) in place for clang 18. Shouldn't this be backported first before adding clang 19 support?

You should feel free to open a PR to upstream your patch.  I'm not seeing a reason to order the changes.

**Comment by MichaelSimons** [comment]
@jkoritzinsky, what is the runtime's process for merging infra changes like this?  Does this require tactics approval or tell mode?

**Comment by jkoritzinsky** [comment]
We usually do tell-mode for infra-only changes. I'm currently OOF, @hoyosjs can you help get this in?

**Comment by hoyosjs** [comment]
Is this release 8 only?

**Comment by hoyosjs** [comment]
Changing to staging branch seems to have messed this. I'll fix this tomorrow.

**Comment by am11** [comment]
> Changing to staging branch seems to have messed this. I'll fix this tomorrow.

Usually infra (build related) changes are directly pushed to the release branch (https://github.com/dotnet/runtime/pull/107921), while product changes go through staging. It's a bit confusing where we draw the line.

**Comment by hoyosjs** [comment]
That's usually only true if there's an undergoing release build and the error is seen after merge of staging, since there won't be more merges.

**Review by jeffschwMSFT — APPROVED** [review]
approved. we can take this as tell mode

cc @carlossanlop

**Comment by MichaelSimons** [comment]
Can someone help get this merged in?  Source-build CI needs this fix.

**Pull request merged**
